juju/juju
# feat: Container manager service

This PR introduces a new way for calculating the value for container networking method. Pre this change there existed state logic that would intercept the model config value on the way into the database and attempt to modify what the user may have set to what this value could be later for the facade.

For example the auto determine value was "", with other possible values being local and provider. If the value set was "" the state layer would stand up an environ and ask if the environ supported container addresses. In turn this would modify the value in model config to reflect provider support.

The problems we have with this approach in that it was only ever done on initial set being either at bootstrap time or model creation. It was never retroactively handled with change to model config. It also modifies the values from the what the user may or may not have set taking away from the user interface aspect of model config.

Essentially we are trying to set in model config what the exact value should be when returned to an agent through the facade instead of just taking the value and making a determination at the facade value. i.e the facade and model config value are highly coupled.

As we move to DQlite we have been looking at this code trying to figure out how to move it over. We initially tried driving the value via model defaults and the provider. The problem with this is we end up in a chicken and egg problem where the provider needs model config to stand up and we need the value from the provider to set model config.

The better approach that this commit aims to start moving towards is to have the facade and service layers interrupt the value on read and translate this into what the facade caller needs.

With this approach we de-couple model config from the facade for the agent. We don't need to do expensive provider stuff for every delta on model config. Lastly we get rid of special chicken and egg problems that have been danced around in the state layer for years.

## Changes

Add a new `agentprovisioner` service with methods
```go
ContainerManagerConfigForType(context.Context, instance.ContainerType) (containermanager.Config, error)
ContainerConfig(ctx context.Context) (params.ContainerConfig, error)
ContainerNetworkingMethod(ctx context.Context) (containermanager.NetworkingMethod, error)
```

The `ContainerNetworkingMethod` replaces `state.Model.AutoConfigureContainerNetworking`, which is now deleted. In model config, `container-networking-method` can be explicitly set to `provider` or `local`, or left empty, in which case the agent provisioner service will automatically determine the container networking method.

The `ContainerManagerConfigForType` and `ContainerConfig` now contain most of the business logic previously in the provisioner facade methods `ContainerManagerConfig` and `ContainerConfig` respectively. These facade methods now outsource most of the business logic to the agent provisioner service. The exception is getting authorised keys for the container manager - we have added a method `GetInitialAuthorisedKeysForContainer` to the key updater service in order to do this. Naturally the `ProvisionerAPI` now depends on `AgentProvisionerService` and `KeyUpdaterService`.

We have added validation checks to model config that:
- on set, `container-networking-method` must be a valid value (`provider`, `local`, or empty string)
- on update, `container-networking-method` cannot be changed.

Minor changes:
- Add new `AgentProvisioner` and `KeyUpdater` methods to the model service factory.
- Remove the old provisioner tests testing `ContainerManagerConfig` and `ContainerConfig`, as they are testing logic that is no longer in the provisioner. Replace them with new tests `provisionerMockSuite.TestContainerManagerConfig` and `.TestContainerConfig` using mocked services.
- Move all provisioner services to a `service.go` file.
- Define some constants for model config keys which didn't have them.
- New type `containermanager.NetworkingMethod` and associated constants should be used everywhere, instead of magic strings "provider" and "local".
- Renaming things, general code quality.

## Checklist


## QA steps



Bootstrap Juju on aws:
```
$ juju bootstrap aws aws
$ juju add-model m
$ juju add-machine
created machine 0
$ juju deploy ubuntu --to lxd:0
```
Check that ubuntu comes up correctly and doesn't crash the container or the host machine.

## Links



**Jira card:** JUJU-6226
